sql: support vector indexes with external row data

DrewKimball opened the pull request:
#### cspann: implement "read-only" mode

This commit adds support for read-only indexes to the cspann library.
If `readOnly=true` is supplied to `cspann.NewIndex`, fixups will be
disabled for that index instance, as will merging into the global
index stats.

Informs #144679

Release note: None

#### sql: support vector indexes with external row data

This commit fixes a few cases where vector-indexing logic was not
correctly using the table's `ExternalRowData` if it was set. It also
initializes `cspann.VectorIndex` instances in read-only mode if the
table with the index has external row data.

Fixes #144679

Release note (bug fix): Fixed a bug in alpha versions of 25.2 as well
as `v25.2.0-beta.1` that would cause vector indexes to return incorrect
or no results from a standby reader in a PCR setup.

DrewKimball committed:
cspann: implement "read-only" mode

This commit adds support for read-only indexes to the cspann library.
If `readOnly=true` is supplied to `cspann.NewIndex`, fixups will be
disabled for that index instance, as will merging into the global
index stats.

Informs #144679

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144681)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144681)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/vecindex/cspann/fixup_processor.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8AaRW6LpBqhIPPIu9:-OO8AaRXDwSQ0C-g1h6y:bkud9uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/fixup_processor.go#L142)):*
> ```Go
> 	// readOnly is true if the index is limited to read-only operations. If it is
> 	// set, fixups are discarded.
> 	readOnly bool
> ```

We shouldn't need a separate field here, instead use `fp.index.options.ReadOnly` to access it.
___
*[`pkg/sql/vecindex/cspann/index_stats.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8Ay-U-Ffj96hpu_cM:-OO8Ay-U-Ffj96hpu_cN:b-welihh) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/index_stats.go#L42)):*
> ```Go
> 	mergeStats mergeStatsFunc
> 
> 	// skipMerge is true if the stats manager should not merge local statistics
> ```

We only update stats for write operations, never for read-only operations. As long as the index is in read-only mode, we'll never try to update stats. I think it would be more intuitive if instead of adding a `skipMerge` flag here, we raise an error (probably an assertion?) if an attempt is made to call any of the mutation methods on `Index`, e.g. `Insert`, `SearchForInsert`, `Delete`, etc.
___
*[`pkg/sql/vecindex/cspann/index.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8AIUZ3yYYAywrvsli:-OO8AIUZ3yYYAywrvslj:b-h67j5o) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/index.go#L199)):*
> ```Go
> 	options *IndexOptions,
> 	stopper *stop.Stopper,
> 	readOnly bool,
> ```

I think "readOnly" should be on IndexOptions - all options should go there.
___
*[`pkg/sql/vecindex/vecstore/store.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8Cq7FAtKGuaB7XUY7:-OO8Cq7FAtKGuaB7XUY8:b-sdiweh) ([raw file](https://github.com/cockroachdb/cockroach/blob/121b2d7a267dcedb5792c8778ce474b88bd8cf90/pkg/sql/vecindex/vecstore/store.go#L149)):*
> ```Go
> }
> 
> // ReadOnly is part of the cspann.Store interface. It returns true if the store
> ```

I think this comment is wrong, correct? It shouldn't be part of the interface.


<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: support vector indexes with external row data

This commit fixes a few cases where vector-indexing logic was not
correctly using the table's `ExternalRowData` if it was set. It also
initializes `cspann.VectorIndex` instances in read-only mode if the
table with the index has external row data.

Fixes #144679

Release note (bug fix): Fixed a bug in alpha versions of 25.2 as well
as `v25.2.0-beta.1` that would cause vector indexes to return incorrect
or no results from a standby reader in a PCR setup.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144681)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @mw5h)

___
*[`pkg/sql/vecindex/cspann/fixup_processor.go` line 142 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8AaRW6LpBqhIPPIu9:-OOR1Gll2QlhgABzm9vX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/fixup_processor.go#L142)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We shouldn't need a separate field here, instead use `fp.index.options.ReadOnly` to access it.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/cspann/index.go` line 199 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8AIUZ3yYYAywrvsli:-OOR7Rp5ADel0M8qbXB3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/index.go#L199)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think "readOnly" should be on IndexOptions - all options should go there.
</blockquote></details>

Done.
___
*[`pkg/sql/vecindex/vecstore/store.go` line 149 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8Cq7FAtKGuaB7XUY7:-OOR0r8W9vc4KoZWrUd3:bilpl5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/121b2d7a267dcedb5792c8778ce474b88bd8cf90/pkg/sql/vecindex/vecstore/store.go#L149)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think this comment is wrong, correct? It shouldn't be part of the interface.
</blockquote></details>

Done. Yes, just left over from an earlier iteration.
___
*[`pkg/sql/vecindex/cspann/index_stats.go` line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144681#-OO8Ay-U-Ffj96hpu_cM:-OOR0fXz2h8U_KKH1wl5:bj3ybec) ([raw file](https://github.com/cockroachdb/cockroach/blob/01c47ecb47f985ed714433dc0ffaae56a620c764/pkg/sql/vecindex/cspann/index_stats.go#L42)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We only update stats for write operations, never for read-only operations. As long as the index is in read-only mode, we'll never try to update stats. I think it would be more intuitive if instead of adding a `skipMerge` flag here, we raise an error (probably an assertion?) if an attempt is made to call any of the mutation methods on `Index`, e.g. `Insert`, `SearchForInsert`, `Delete`, etc.
</blockquote></details>

Done. I already added assertions to the store, so I've just removed the plumbing from the index stats logic.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144681)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTR!

bors r=andy-kimball
